guide/src/class/protocols.md

```diff
@@ -147,6 +141,108 @@ Other:
   * `fn __index__(&'p self) -> PyResult<impl ToPyObject>`
   * `fn __round__(&'p self, ndigits: Option<impl FromPyObject>) -> PyResult<impl ToPyObject>`
 
+### Emulating sequential containers (such as lists or tuples)
+
+The [`PySequenceProtocol`] trait allows to emulate
+[sequential container types](https://docs.python.org/3/reference/datamodel.html#emulating-container-types).
+
+For a sequence, the allowable keys should be the integers _k_ for which _0 <= k < N_,
+where _N_ is the length of the sequence.
+
+  * `fn __len__(&self) -> PyResult<usize>`
+
+    Implements the built-in function `len()` for the sequence.
+
+  * `fn __getitem__(&self, idx: isize) -> PyResult<impl ToPyObject>`
+
+    Implements evaluation of the `self[idx]` element.
+    If the `idx` value is outside the set of indexes for the sequence, `IndexError` should be raised.
+
+    *Note:* Negative integer indexes are handled as follows: if `__len__()` is defined,
+    it is called and the sequence length is used to compute a positive index,
+    which is passed to `__getitem__()`.
+    If `__len__()` is not defined, the index is passed as is to the function.
```

Comment:
This differs from https://docs.python.org/3/reference/datamodel.html#object.__getitem__, which says that negative integers are are passed directly to `__getitem__` which then can interpret the negative numbers as it likes?

Reply:
I presume PyO3 plugs the `__getitem__` method implementation into the [`PySequenceMethods.sq_item`](https://docs.python.org/3/c-api/typeobj.html#c.PySequenceMethods) slot. The C function in that slot has a slightly different interface from the Python-native `__getitem__()` method, as described in the C-API docs linked above. `PyObject_GetItem()` and `PySequence_GetItem()` handle the negative indexes themselves when `__len__()` is also implemented.

My implementations of `__getitem__()` always start with `assert!(idx >= 0);` because I also implement `__len__()`, and they indeed work with the negative indexes.

Reply:
Ah yes, this is a really good point. Perhaps in the documentation here we should link all these methods to their relevant slot information?

All the relationships between these methods and their slots is defined in https://github.com/PyO3/pyo3/blob/main/pyo3-macros-backend/src/defs.rs

Reply:
> Perhaps in the documentation here we should link all these methods to their relevant slot information?

I like this idea, but the mapping between methods and slots looks not so straightforward. There are methods that do not have a C function slot like `__bytes__()`, `__format__()` and `__reversed__()`. And then there are methods that share the same slot like `__setitem__()` and `__delitem__()`. I think we could add a symbolic tag like `PySequenceMethods.sq_item` to each trait method.

Another interesting issue here is the implicit slot method probing order: for example the evaluation of `obj[i]` Python expression probes `PyMappingMethods.mp_subscript` before `PySequenceMethods.sq_item`. This is relevant when someone implements both of these traits for some reason.